### DISCUSSION TITLE
Write server actions once - call it two ways: from server and from client wrapped in next-safe-action

### DISCUSSION THREAD

**Discussion by sp88011** [next-safe-action/next-safe-action#65] [Q&A]
I realize that `next-safe-actions` is meant to provide a way to call server actions **from the client**. 

I'm looking for ideas for how to structure my code such that I only have to write a server function **once**, and then be able to call it from both the server and from the client using next-safe-action.

I realize I can call a classic server action (not wrapped in `next-safe-action`) both ways, from the server and from the client. But I would like to keep all the benefits of next-safe-actions (e.g. loading states, callbacks)...

It feels wrong to write every function twice: Example

```
//app/_actions.ts

"use server"

// this is the normal server action
export async function callMeMaybe (phone:string) {
 await new Promise((resolve) => setTimeout(resolve, 100));
 return `${phone} did not pick up`.
}


//and now wrapped in next-safe-action

const schema = z.object({
 phone: z.string(),
});

export const callMeMaybeAction ( schema, async ( {phone} ) => {
 await new Promise((resolve) => setTimeout(resolve, 100));
 return `${phone} did not pick up`.
})
````

**Comment by TheEdoRan** [comment]
> I'm looking for ideas for how to structure my code such that I only have to write a server function once, and then be able to call it from both the server and from the client using next-safe-action.

Actions defined using next-safe-action are just regular functions with input validation. They can be called from the server too, obviously without using hooks. Have you found problems while doing so?
